File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        // 예외 처리
        if(root == null) {
            return null;
        }

        // 반복 BFS 처리를 위한 큐 선언
        Queue<TreeNode> q = new LinkedList<>();

        // 큐에 루트부터 삽입
        q.add(root);

        // 직렬화 결과 변수 - 리턴 타입이 문자열에 계속 원소가 추가되는 형태이므로 StringBuilder로 선언한다.
        StringBuilder sb = new StringBuilder();

        // 0번 인덱스는 사용하지 않는다. 초기 값은 루트 노드의 값
        // 이때 사용하지 않는 인덱스거나 노드가 없는 인덱스(null)은 편의상 #(사용하지 않는 문자)로 표현한다.
        sb.append(String.valueOf("#," + root.val));

        // BFS 탐색
        while(!q.isEmpty()) {
            // 큐에서 노드 추출
            TreeNode node = q.remove();

            // 왼쪽 자식 노드가 null이 아니면 큐와 직렬화 결과 문자열에 삽입
            if(node.left != null) {
                q.add(node.left);
                sb.append("," + node.left.val);
            } else {
                // 왼쪽 자식 노드가 null이면 직렬화 결과에 # 삽입
                sb.append(",#");
            }

            // 오른쪽 자식 노드가 null이 아니면 큐와 직렬화 결과 문자열에 삽입
            if(node.right != null) {
                q.add(node.right);
                sb.append("," + node.right.val);
            } else {
                // 오른쪽 자식 노드가 null이면 직렬화 결과에 # 삽입
                sb.append(",#");
            }
        }

        // StringBuilder를 문자열로 변환하여 반환
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        // 예외 처리
        if(data == null) {
            return null;
        }
        
        // 콤마(,)를 기준으로 문자열 분리
        String[] nodes = data.split(",");
        
        // 인덱스 1번이 루트
        TreeNode root = new TreeNode(Integer.parseInt(nodes[1]));

        // 직렬화와 동일하게 BFS로 역직렬화 처리 위한 큐 선언
        Queue<TreeNode> q = new LinkedList<>();

        // 큐에 루트부터 추가
        q.add(root);

        // 루트를 제외하고 2번 인덱스부터 진행
        int index = 2;

        while(!q.isEmpty()) {
            TreeNode node = q.remove();

            String val = nodes[index];

            // 해당 인덱스가 #이 아니면 왼쪽 자식 노드로 설정하고 큐에 추가
            if(!nodes[index].equals("#")) {
                node.left = new TreeNode(Integer.parseInt(nodes[index]));
                q.add(node.left);
            }

            // 항상 좌/우 세트로 구성되므로 한 번 더 진행
            index += 1;
            if(!nodes[index].equals("#")) {
                node.right = new TreeNode(Integer.parseInt(nodes[index]));
                q.add(node.right);
            }

            index += 1;
        }

        // 결과 반환
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
	// Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        // 예외 처리
        if(root == null) {
            // serialize() 메서드의 반환형이 String (null 불가)
            return ""
        }
        
        // BFS 탐색을 위한 큐 선언
        val q: Queue<TreeNode> = LinkedList()
        
        // 직렬화 결과를 저장할 문자열 변수
        val sb = StringBuilder()
        
        // root 노드를 큐에 추가
        q.add(root)
        
        // 0번째 인덱스는 사용하지 않으며("#") 다음 인덱스는 root부터 시작
        sb.append("#," + root.`val`)
        
        // BFS 탐색하며 트리를 문자열로 직렬화
        while(!q.isEmpty()) {
            // 큐에서 노드 추출
            val node = q.remove()
            
            // 왼쪽 자식 노드가 null이 아니면 q에 추가하고 sb에 값을 추가(append)
            if(node.left != null) {
                q.add(node.left)
                sb.append("," + node.left.`val`)
            } else {
                // null이라면 sb에 사용하지 않는 문자("#")로 null 임을 표시
                sb.append(",#")
            }
            
            // 오른쪽 자식 노드가 null이 아니면 q에 추가하고 sb에 값을 추가(append)
            if(node.right != null) {
                q.add(node.right)
                sb.append("," + node.right.`val`)
            } else {
                // null이라면 sb에 사용하지 않는 문자("#")로 null 임을 표시
                sb.append(",#")
            }
        }
        
        // 문자열로 변환하여 반환
        return sb.toString()
            
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        // 예외 처리
        if(data == "") {
            // deserialize() 메서드의 반환형이 TreeNode? (null 가능)
            return null
        }
        
        // data 문자열을 ,를 기준으로 분리
        val nodes = data.split(",")
        
        // 인덱스 1번이 root 노드
        val root = TreeNode(nodes[1].toInt())
        
        // BFS 탐색을 위한 큐 선언
        val q: Queue<TreeNode> = LinkedList()
        
        // 첫번째 원소로 root 노드 추가
        q.add(root)
        
        // 2번 인덱스부터 탐색 시작
        var index = 2
        
        // BFS 탐색 시작
        while(!q.isEmpty()) {
            val node = q.remove()
            
            // 값이 있다면 왼쪽 자식으로 추가
            if(nodes[index] != "#") {
                node.left = TreeNode(nodes[index].toInt())
                
                // BFS 탐색을 위해 큐에 추가
                q.add(node.left)
            }
            
            // 항상 좌/우 세트로 구성되어있으므로 오른쪽 자식 노드에 대해서도 한번 더 진행
            index += 1
            if(nodes[index] != "#") {
                node.right = TreeNode(nodes[index].toInt())
                
                // BFS 탐색을 위해 큐에 추가
                q.add(node.right)
            }
            
            // 다음 인덱스에 대해 탐색 진행
            index += 1
        }
        
        
        // 결과 노드 반환
        return root
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * var ser = Codec()
 * var deser = Codec()
 * var data = ser.serialize(longUrl)
 * var ans = deser.deserialize(data)
 */